core/src/main/java/org/elasticsearch/cockroach/CockroachActionExecutor.java

```diff
@@ -0,0 +1,363 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.cockroach;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.cluster.ClusterChangedEvent;
+import org.elasticsearch.cluster.ClusterStateListener;
+import org.elasticsearch.cockroach.CockroachActionRegistry.CockroachActionHolder;
+import org.elasticsearch.cockroach.CockroachTasksInProgress.CockroachTaskInProgress;
+import org.elasticsearch.common.Nullable;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.component.AbstractComponent;
+import org.elasticsearch.common.inject.Provider;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.util.concurrent.AbstractRunnable;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.tasks.TaskManager;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportResponse.Empty;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * This component is responsible for execution of cockroach actions on individual nodes. It runs on all
+ * non-transport client nodes in the cluster and monitors cluster state changes to detect started commands.
+ */
+public class CockroachActionExecutor extends AbstractComponent implements ClusterStateListener {
+    private final Map<String, RunningCockroachTask> runningTasks = new HashMap<>();
+    private final CockroachActionRegistry cockroachActionRegistry;
+    private final TaskManager taskManager;
+    private final ThreadPool threadPool;
+    private final CockroachTransportService cockroachTransportService;
+
+
+    public CockroachActionExecutor(final Settings settings, final CockroachActionRegistry cockroachActionRegistry,
+                                   final TaskManager taskManager, final CockroachTransportService cockroachTransportService,
+                                   final ThreadPool threadPool) {
+        super(settings);
+        this.cockroachActionRegistry = cockroachActionRegistry;
+        this.taskManager = taskManager;
+        this.cockroachTransportService = cockroachTransportService;
+        this.threadPool = threadPool;
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        CockroachTasksInProgress tasks = event.state().custom(CockroachTasksInProgress.TYPE);
+        CockroachTasksInProgress previousTasks = event.previousState().custom(CockroachTasksInProgress.TYPE);
+
+        if (Objects.equals(tasks, previousTasks) == false || event.nodesChanged()) {
```

Comment:
I'm wondering if we need to use `Objects.equals` here which would be quite heavy-weight on the entire CockroachTasksInProgress... in this case, for example, all we care about is if there are new task entries whose executor is the local node id... in that case, maybe we can have a method on `CockroachTasksInProgress` such as `hasNewEntriesForExecutor(localNodeId)`